wdl-doc/src/docs_tree.rs

```diff
@@ -0,0 +1,336 @@
+//! Implementations for a [`DocsTree`] which represents the DOCS directory.
+
+use std::collections::BTreeMap;
+use std::fs::canonicalize;
+use std::path::Path;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+use maud::Markup;
+use maud::html;
+use pathdiff::diff_paths;
+
+use crate::Document;
+use crate::full_page;
+use crate::r#struct::Struct;
+use crate::task::Task;
+use crate::workflow::Workflow;
+
+/// The type of a page.
+#[derive(Debug)]
+pub enum PageType {
+    /// An index page.
+    Index(Document),
+    /// A struct page.
+    Struct(Struct),
+    /// A task page.
+    Task(Task),
+    /// A workflow page.
+    Workflow(Workflow),
+}
+
+/// An HTML page in the DOCS directory.
+#[derive(Debug)]
+pub struct HTMLPage {
+    /// The display name of the page.
+    name: String,
+    /// The type of the page.
+    page_type: PageType,
+}
+
+impl HTMLPage {
+    /// Create a new HTML page.
+    pub fn new(name: String, page_type: PageType) -> Self {
+        Self { name, page_type }
+    }
+
+    /// Get the name of the page.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Get the type of the page.
+    pub fn page_type(&self) -> &PageType {
+        &self.page_type
+    }
+}
+
+/// A node in the DOCS directory tree.
+#[derive(Debug)]
+struct Node {
+    /// The name of the node.
+    name: String,
+    /// The absolute path to the node.
+    path: PathBuf,
+    /// The page associated with the node.
+    page: Option<Rc<HTMLPage>>,
+    /// The children of the node.
+    children: BTreeMap<String, Node>,
+}
+
+impl Node {
+    /// Create a new node.
+    pub fn new<P: Into<PathBuf>>(name: String, path: P) -> Self {
+        Self {
+            name,
+            path: path.into(),
+            page: None,
+            children: BTreeMap::new(),
+        }
+    }
+
+    /// Get the name of the node.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Get the path of the node.
+    pub fn path(&self) -> &PathBuf {
+        &self.path
+    }
+
+    /// Get the page associated with the node.
+    pub fn page(&self) -> Option<Rc<HTMLPage>> {
+        self.page.clone()
+    }
+
+    /// Iterate over the children of the node in a Depth First Traversal order.
+    pub fn depth_first_traversal(&self) -> Vec<&Node> {
+        let mut nodes = Vec::new();
+        nodes.push(self);
+
+        for child in self.children.values() {
+            nodes.extend(child.depth_first_traversal());
+        }
+
+        nodes
+    }
+}
+
+/// A tree representing the DOCS directory.
+#[derive(Debug)]
+pub struct DocsTree {
+    /// The root of the tree.
+    ///
+    /// `root.path` is the path to the DOCS directory and is absolute.
+    root: Node,
+    /// The absolute path to the stylesheet, if it exists.
+    stylesheet: Option<PathBuf>,
+}
+
+impl DocsTree {
+    /// Create a new DOCS tree.
+    pub fn new<P: AsRef<Path>>(root: P, stylesheet_to_copy: Option<P>) -> Self {
```

Comment:
When repeating a type parameter in a signature like this, it's probably best to use `impl AsRef<Path>` in the signatures so that they don't have to be the same type (i.e. one should be able to call this with `new("foo", Some(Path::new("foo")))`. It's functionally equivalent to having two type parameters `P: AsRef<Path>, Q: AsRef<Path>`.

However, the `Option<impl AsRef<Path>>` is annoying to deal with as you would have to qualify it in the `None` case.

When I see something like that, I tend to break the method into two: `new(root: impl AsRef<Path>)` and `new_with_stylesheet(root: impl AsRef<Path>, stylesheet: impl AsRef<Path>)` and let it internally deal with the `Option`.

Reply:
Would ask over Slack, but it's still down! I've updated `DocsTree::new` to just take an `impl AsRef<Path>` and then wrote `new_with_stylesheet` like you usggested.

The pattern you pointed here is also in use for the `document_workspace()` entrypoint function, but I'm not sure the right way to refactor that. Should there be a pair of `document_worskspace()` and `document_workspace_with_stylesheet()` functions? Or is there another way to refactor it that you would recommend?

Reply:
`document_workspace_with_stylesheet` definitely is a mouthful; if we want two different functions, I'm sure we'd all prefer something less verbose.

Still, perhaps having the `Option` parameter is fine and folks won't mind passing something like `None::<&str>`? I don't have any suggested alternatives, but I still think we should be using different type parameters at the very least.